OverlayPlugin.Core/NetworkProcessors/LineAbilityExtra.cs

```diff
@@ -0,0 +1,123 @@
+        public LineAbilityExtra(TinyIoCContainer container)
+        {
+            logger = container.Resolve<ILogger>();
+            ffxiv = container.Resolve<FFXIVRepository>();
+            var netHelper = container.Resolve<NetworkParser>();
+            if (!ffxiv.IsFFXIVPluginPresent())
+                return;
+            try
+            {
+                var mach = Assembly.Load("Machina.FFXIV");
+                var msgHeaderType = mach.GetType("Machina.FFXIV.Headers.Server_MessageHeader");
+                actionEffectHeaderType = mach.GetType("Machina.FFXIV.Headers.Server_ActionEffectHeader");
+                fieldR = actionEffectHeaderType.GetField("rotation");
+
+                int[] sizes = {1, 8, 16, 24, 32};
+                foreach (int size in sizes)
+                {
+                    Type aeType = mach.GetType("Machina.FFXIV.Headers.Server_ActionEffect" + size);
+                    var actionEffectTypeInfo = new ActionEffectTypeInfo()
+                    {
+                        machinaType = aeType,
+                        packetSize = Marshal.SizeOf(aeType)
+                    };
+                    minSize = Math.Min(actionEffectTypeInfo.packetSize, minSize);
+                    opcodeToType.Add(netHelper.GetOpcode("Ability" + size), actionEffectTypeInfo);
+                }
+
+                offsetMessageType = netHelper.GetOffset(msgHeaderType, "MessageType");
+                ffxiv.RegisterNetworkParser(MessageReceived);
+            }
+            catch (System.IO.FileNotFoundException)
+            {
+                logger.Log(LogLevel.Error, Resources.NetworkParserNoFfxiv);
+            }
+            catch (Exception e)
+            {
+                logger.Log(LogLevel.Error, Resources.NetworkParserInitException, e);
+            }
+
+            var customLogLines = container.Resolve<FFXIVCustomLogLines>();
+            this.logWriter = customLogLines.RegisterCustomLogLine(new LogLineRegistryEntry()
+            {
+                Name = "ActorCastExtra",
+                Source = "OverlayPlugin",
+                ID = LogFileLineID,
+                Version = 1,
+            });
+        }
+
+        private unsafe void MessageReceived(string id, long epoch, byte[] message)
+        {
+            if (message.Length < minSize)
+                return;
+
+            fixed (byte* buffer = message)
+            {
+                int opCode = *(ushort*)&buffer[offsetMessageType];
+                var info = opcodeToType[opCode];
+                if (info != null)
+                {
+                    if (message.Length < info.packetSize)
+                    {
+                        return;
+                    }
+                    object packet = Marshal.PtrToStructure(new IntPtr(buffer), actionEffectHeaderType);
+                    DateTime serverTime = ffxiv.EpochToDateTime(epoch);
+                    float x = (BitConverter.ToUInt16(message, message.Length - 12) - 0x7FFF) / 32.767f;
```

Comment:
Rather than repeating this logic, it would be better to extract out a helper function e.g. `float x = UShortToFloatHelper(BitConverter.ToUInt16(message, message.Length - 12))`

This can exist in FFXIVRepository alongside `EpochToDateTime`

Reply:
Done